harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,15 +34,138 @@
     pass
 
 
-class PyTorchTrialController(det.TrialController):
-    def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)
+class TrainUnit:
+    def __init__(self, value):
+        self.value = value
+
+    @staticmethod
+    def from_searcher_unit(length: int, unit: core.Unit):
+        if unit == core.Unit.EPOCHS:
+            return Epoch(length)
+        elif unit == core.Unit.RECORDS:
+            return Record(length)
+        elif unit == core.Unit.BATCHES:
+            return Batch(length)
+        else:
+            raise ValueError(f"unrecognized searcher unit {unit}")
+
+    @staticmethod
+    def from_values(batches: Optional[int] = None,
+                    records: Optional[int] = None,
+                    epochs: Optional[int] = None):
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches:
+            return Batch(batches)
+        elif records:
+            return Record(records)
+        elif epochs:
+            return Epoch(epochs)
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+class Record(TrainUnit):
+    pass
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level train loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+
+class TrialState:
+    def __init__(
+        self,
+        trial_id: int = 0,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
+        step_id: int = 0,
+        last_val: int = 0,
+        batches_trained: int = None,
+        epochs_trained: int = None,
+    ) -> None:
+        # Store TrialID to distinguish between e.g. pause/restart and continue training.
+        self.trial_id = trial_id
+        self.last_ckpt = last_ckpt
+        self.steps_completed = steps_completed
+        self.step_id = step_id
+        self.last_val = last_val
+        self.batches_trained = batches_trained
+        self.epochs_trained = epochs_trained
+
+
+class PyTorchTrialController:
+    def __init__(self, trial_inst: det.Trial,
+                 context: pytorch.PyTorchTrialContext,
+                 min_checkpoint_period: TrainUnit,
+                 min_validation_period: TrainUnit,
+                 average_training_metrics: Optional[bool] = True,
+                 smaller_is_better: Optional[bool] = True,
+                 max_length: Optional[TrainUnit] = None,
+                 steps_completed: Optional[int] = 0,
+                 latest_checkpoint: Optional[str] = None,
+                 local_training: Optional[bool] = False,
```

Comment:
why does the user need to specify `local_training`?  It feels more like a footgun, if they accidentally have it set to `True` when they submit their script.